pkg/auth/mtls_authhandler.go

```diff
@@ -0,0 +1,279 @@
+	}
+
+	caBundle, err := m.cert.GetTrustBundle()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get CA bundle: %w", err)
+	}
+
+	// set up TCP connection
+	conn, err := net.Dial("tcp", net.JoinHostPort(ar.remoteHostIP.String(), strconv.Itoa(m.cfg.MTLSListenerPort)))
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial %s:%d: %w", ar.remoteHostIP.String(), m.cfg.MTLSListenerPort, err)
+	}
+	defer conn.Close()
+
+	var expirationTime *time.Time = &clientCert.Leaf.NotAfter
+
+	// set up TLS socket
+
+	//nolint:gosec // InsecureSkipVerify is not insecure as we do the verification in VerifyPeerCertificate
+	tlsConn := tls.Client(conn, &tls.Config{
+		ServerName: m.cert.NumericIdentityToSNI(ar.remoteIdentity),
+		GetClientCertificate: func(info *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			return clientCert, nil
+		},
+		MinVersion:         tls.VersionTLS13,
+		InsecureSkipVerify: true, // not insecure as we do the verification in VerifyPeerCertificate
+		VerifyPeerCertificate: func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			// verifiedChains will be nil as we set InsecureSkipVerify to true
+
+			chain := make([]*x509.Certificate, len(rawCerts))
+			for i, rawCert := range rawCerts {
+				cert, err := x509.ParseCertificate(rawCert)
+				if err != nil {
+					return fmt.Errorf("failed to parse certificate: %w", err)
+				}
+				chain[i] = cert
+			}
+
+			peerExpirationTime, err := m.verifyPeerCertificate(&ar.remoteIdentity, caBundle, [][]*x509.Certificate{chain})
+			if peerExpirationTime != nil && peerExpirationTime.Before(*expirationTime) {
+				expirationTime = peerExpirationTime // send down the lowest expiration time of the two certificates
+			}
+			return err
+		},
+		ClientCAs: caBundle,
+		RootCAs:   caBundle,
+	})
+	defer tlsConn.Close()
+
+	if err := tlsConn.Handshake(); err != nil {
+		return nil, fmt.Errorf("failed to perform TLS handshake: %w", err)
+	}
+
+	if expirationTime == nil {
+		return nil, fmt.Errorf("failed to get expiration time of peer certificate")
+	}
+
+	return &authResponse{
+		expirationTime: *expirationTime,
+	}, nil
+}
+
+func (m *mtlsAuthHandler) authType() policy.AuthType {
+	return policy.AuthTypeMTLSSpiffe
+}
+
+func (m *mtlsAuthHandler) listenForConnections(upstreamCtx context.Context, ready chan<- struct{}) {
+	// set up TCP listener
+
+	ctx, cancel := context.WithCancel(upstreamCtx)
+	defer cancel()
+
+	var lc net.ListenConfig
+	l, err := lc.Listen(ctx, "tcp", fmt.Sprintf(":%d", m.cfg.MTLSListenerPort))
+	if err != nil {
+		m.log.WithError(err).Fatal("Failed to start mTLS listener")
+	}
+	go func() { // shutdown socket goroutine
+		<-ctx.Done()
+		l.Close()
+	}()
+
+	m.log.WithField(logfields.Port, m.cfg.MTLSListenerPort).Info("Started mTLS listener")
+	ready <- struct{}{} // signal to hive that we are ready to accept connections
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
```

Comment:
nitpicking: I don't think we need this `select` anymore, as the listener will be closed once the context is done.